test: topology: test concurrent bootstrap

patjed41 opened the pull request:
We add a test for concurrent bootstrap in the raft-based topology.

Additionally, we extend the testing framework with a new function -
`ManagerClient.servers_add`. It allows adding multiple servers
concurrently to a cluster.

This PR is the first step to fix #15423. After merging it, if the new test
doesn't fail for some time in CI, we can:
- use `ManagerClient.servers_add` in other tests wherever possible,
- start initial servers concurrently in all suites with
  `initial_size > 0`.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4772/
- Duration: 2 hr 36 min
- Builder: spider3.cloudius-systems.com

kbr-scylla commented on test/pylib/manager_client.py:
```diff
@@ -229,6 +229,41 @@ async def server_add(self, replace_cfg: Optional[ReplaceConfig] = None,
+            server_infos = await self.client.put_json("/cluster/addservers", data, response_type="json",
+                                                      timeout=ScyllaServer.TOPOLOGY_TIMEOUT * servers_num)
+        except Exception as exc:
+            raise Exception("Failed to add servers") from exc
```
Please verify if this preserves the original contents of `exc` (e.g. by modifying Scylla to throw on boot and seeing what happens). I recall having some problems with this `from exc` thing, where details about the error were lost and made debugging failures from CI harder; formatting the original as string and including it into the new exception's string gave me better results (maybe we could/should do both?)

patjed41 replied:
It doesn't preserve the original contents of `exc` and I think it works as intended. This `from exc` thing is just to indicate in code that the raised exception is a direct consequence of the catched exception.

What do you mean by "formatting the original as string"? If I simply print `exc`, I get:
```
        except Exception as exc:
>           raise Exception(f"Failed to add servers: {exc}") from exc
E           Exception: Failed to add servers: HTTP error 500, uri: http+unix://api/cluster/addservers, params: None, json: {'start': True, 'servers_num': 3}, body:
E           SOME ERROR
```
I don't think we want to show this whole HTTP info. We already log it in `RESTClient._fetch`. I don't think logging it again makes a lot of sense.

Another option is to catch `HTTPError` instead and print `exc.message`. But again, it will be logged for the second time (maybe in a nicer way though). Also, it might not be correct if, for example, a `RuntimeError` raise is also possible.

Maybe it's better to leave it as it is? If you still prefer printing `exc`, please specify how. Then, what should I do with the rest of the `from exc` things present in the code? I could change all of them in the same way in a follow-up.

kbr-scylla replied:
I meant `str(exc)` (which is also what `f"... {exc} ..."` does IIRC).

The `SOME ERROR` part of the message contains the root cause of the error and it should definitely be shown when test fails. The HTTP thing is not necessary I agree, but it also doesn't really hurt.

If it's as you said, that we're already printing this error earlier, then it's fine. What's important is that we can easily find it in the logs.

kbr-scylla commented on test/pylib/scylla_cluster.py:
```diff
@@ -760,6 +768,22 @@ async def add_server(self, replace_cfg: Optional[ReplaceConfig] = None,
+
+        # If the cluster is empty, all servers must use the same seed to not start separate clusters.
+        if not self.servers:
+            self.initial_seed = await self._get_new_ip()
```
This is leaking IPs if `add_server` below fails, no?

kbr-scylla replied:
Also, instead of having this `is_initial_seed` parameter, wouldn't it be more elegant to lease all IPs up-front here, and pass them to `add_server`? `add_server` would take an IP address as a parameter, forcing the caller to provide IP. Perhaps `add_server` could be turned into a private function `_add_server`, and there would be a simple `add_server` wrapper that obtains the IP and calls `_add_server`. `add_servers` would similarly call `_add_server`. The wrappers would release IP(s) on error.

patjed41 replied:
I solved both of these problems by moving `initial_seed` initialization to `add_server`.

But I've just realized that introducing `_get_new_ip` is pointless now since it is and will be used only once. I'll quickly resubmit the PR.

kostja commented on service/storage_service.cc:
```diff
@@ -6656,6 +6656,15 @@ future<join_node_response_result> storage_service::join_node_response_handler(jo
 
     co_return co_await std::visit(overloaded_functor {
         [&] (const join_node_response_params::accepted& acc) -> future<join_node_response_result> {
+            // Allow other nodes to mark the replacing node as alive. It has
```
I don't understand this comment. Is it one more place where we have to trust gossip UP/DOWN states because we have no tracking of UP/DOWN states of our own? @gleb-cloudius 

What problem are we solving here?

kostja replied:
I disagree with the purpose: why the heck we need to even look at the gossip alive state of a node when assembling the cluster?

patjed41 replied:
> I don't understand this comment. Is it one more place where we have to trust gossip UP/DOWN states because we have no tracking of UP/DOWN states of our own? @gleb-cloudius
>
> What problem are we solving here?

This commit belongs to #15926. I mentioned it in the description.

The problem it solves is described in the commit message:
```
After this change, other nodes can see the replacing node as alive
only after the topology coordinator accepts its join request.

In the following commits, we make the topology coordinator reject
join requests if the node being replaced is considered alive by the
gossiper. However, the replacing node may become alive before the
topology coordinator does the validation. If the replacing node
reuses the IP of the node being replaced, the topology coordinator
cannot know which of these two nodes is alive and whether it should
reject the join request.
```
We don't trust gossip here. It's the other way around -- the gossip state depends on the topology coordinator accepting the join request.

> I disagree with the purpose: why the heck we need to even look at the gossip alive state of a node when assembling the cluster?

We don't need to. In #15926, we reject replace if the node being replaced is considered alive to improve user experience. This change is not entirely safe, but it will be once we implement banning and ban the node being replaced and the ignored nodes at the beginning of replace. It's discussed in https://github.com/scylladb/scylladb/pull/15926#discussion_r1399354469.

gleb-cloudius replied:
> I don't understand this comment. Is it one more place where we have to trust gossip UP/DOWN states because we have no tracking of UP/DOWN states of our own? @gleb-cloudius
> 

I have no idea what *you* mean by "tracking of UP/DOWN states of our own". We have the tracking. It is a gossiper's failure detector that we use here. There also not problem trusting UP state. DOWN state is a different story and this is what we discussed with @kbr-scylla here https://github.com/scylladb/scylladb/pull/15926#discussion_r1399362458

> What problem are we solving here?

The UI problem. Currently it is not allowed to remove a live node (you get `Startup failed: std::runtime_error (Cannot replace a live node...)` and it should stay so. It makes a lot of sense. An operator error should not kill a node.

patjed41 commented:
v2:
- rebased on v4 of #15926, once it is merged I'll rebase this PR again
- moved `initial_seed` initialization to `server_add` and thanks to this the `is_initial_seed` became unneeded
- updated comments to better describe the new way of using `initial_seed`
- removed introduction of `ScyllaCluster_get_new_ip` as this function wouldn't be used more than once anymore

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:x: - Unit Tests
#### Failed Tests (2/21886):
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4863/testReport/junit/boost.cql_query_test.test_clustering_filtering_2.1.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4863/testReport/junit/boost.role_manager_test.create_role.2.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4863/
- Duration: 3 hr 35 min
- Builder: i-00661b14402dfc8cc

patjed41 commented:
> #### 🔴 CI State: FAILURE
> 
> ✅ - Build ✅ - Sanity Tests ❌ - Unit Tests
> #### Failed Tests (2/21886):
> 
>     * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4863/testReport/junit/boost.cql_query_test.test_clustering_filtering_2.1.xunit/xml/Tests___Unit_Tests____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+%5Bempty%5D)
> 
>     * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4863/testReport/junit/boost.role_manager_test.create_role.2.xunit/xml/Tests___Unit_Tests____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+%5Bempty%5D)
> 
> 
> #### Build Details:
> 
>     * Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4863/
> 
>     * Duration: 3 hr 35 min
> 
>     * Builder: i-00661b14402dfc8cc

Caused by #16144

patjed41 commented:
Rebased on master after #15926 got merged and updated PR's description.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4881/
- Duration: 2 hr 16 min
- Builder: spider5.cloudius-systems.com
